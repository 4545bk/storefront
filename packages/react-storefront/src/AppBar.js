/**
 * @license
 * Copyright Â© 2017-2018 Moov Corporation.  All rights reserved.
 */
import React, { Component } from 'react'
import PropTypes from 'prop-types'
import classnames from 'classnames'
import { throttle } from 'lodash'
import Toolbar from '@material-ui/core/Toolbar'
import SvgIcon from '@material-ui/core/SvgIcon'
import { observer, inject } from 'mobx-react'
import Hidden from '@material-ui/core/Hidden'
import withStyles from '@material-ui/core/styles/withStyles'
import { Vbox } from './Box'
import ToolbarButton from './ToolbarButton'
import MenuIcon from './assets/menu.svg'

export const styles = (theme) => ({
  root: {
    boxSizing: 'content-box',
    position: 'relative',
    zIndex: theme.zIndex.modal + 10
  },

  withAmp: {
    zIndex: theme.zIndex.amp.modal + 1
  },

  icon: {
    color: theme.palette.action.active
  },

  toolBar: {
    height: theme.headerHeight || '64px',
    padding: `0 7px`,
    maxWidth: theme.maxWidth,
    flex: 1
  },

  wrap: {
    borderBottom: `1px solid ${theme.palette.divider}`,
    position: 'absolute',
    top: 0,
    left: 0,
    right: 0,
    backgroundColor: theme.palette.background.paper,
    zIndex: theme.zIndex.modal + 10,
    display: 'flex',
    justifyContent: 'center',
  },

  stuck: {
    transform: 'translateY(0)',
  },

  unstuck: {
    transform: 'translateY(-100%)',
  },

  animate: {
    transition: 'transform .15s ease-in',
  },

  hidden: {
    position: 'fixed',
    zIndex: theme.zIndex.modal + 10,
    boxShadow: theme.shadows[2],
    top: 0,
    left: 0,
    right: 0
  },

  menuOpen: {
    boxShadow: 'none'
  },

  link: {
    textDecoration: 'none'
  }
});

/**
 * A header that auto hides when the user scrolls down and auto shows when the user 
 * scrolls up. A hamburger button that controls the menu is automatically displayed on
 * the left side of the header.  Children are placed directly to the right of the menu
 * button.
 */
@inject(({ app }) => ({ menu: app.menu, amp: app.amp }))
@withStyles(styles, { name: 'RSFAppBar'})
@observer
export default class Header extends Component {

  state = {
    stuck: false,
    hidden: false,
    animate: false
  }

  static propTypes = {
    /**
     * CSS classes
     */
    classes: PropTypes.object,

    /**
     * Set to true to show the hamburger button only on mobile devices. Defaults to false.
     */
    responsive: PropTypes.bool,

    /**
     * A component for the menu icon.
     */
    MenuIcon: PropTypes.func
  }

  static defaultProps = {
    responsive: false,
    MenuIcon
  }

  render() {
    const { MenuIcon, classes, children, fixed, menu, responsive, width, amp, showLabels } = this.props

    const menuIcon = <SvgIcon className={classes.icon}><MenuIcon/></SvgIcon>

    return (
      <div className={classnames({ [classes.root]: true, [classes.withAmp]: amp })}> 
        <div className={classnames({
          [classes.wrap]: true,
          [classes.fixed]: fixed,
          [classes.hidden]: this.state.hidden || menu.open,
          [classes.stuck]: (this.state.hidden && this.state.stuck) || menu.open ,
          [classes.unstuck]: this.state.hidden && !this.state.stuck,
          [classes.animate]: this.state.animate,
          [classes.menuOpen]: menu.open
        })}>
          <Toolbar disableGutters classes={{ root: classes.toolBar }}>
            <Hidden mdUp={responsive} implementation="css">
              <a on="tap:moov_menu.toggle" className={classes.link}>
                <ToolbarButton 
                  aria-label="Menu" 
                  color="inherit" 
                  onClick={menu.toggle}
                  icon={menuIcon}
                />
              </a>
            </Hidden>
            { children }
          </Toolbar>
        </div>
      </div>
    )
  }

  componentDidMount() {
    window.addEventListener('scroll', this.onScroll, { passive: true })
  }

  componentWillUnmount() {
    window.removeEventListener('scroll', this.onScroll, { passive: true })
  }

  componentDidUpdate(prevProps, prevState) {
    if (!prevState.hidden && this.state.hidden) {
      setTimeout(() => this.setState({ animate: true }), 100)
    }
  }

  onScroll = () => {
    const height = 64, { scrollY } = window, { lastScrollY } = this
    const { menu } = this.props

    if (this.state.hidden) {
      if (scrollY <= 0) {
        this.setState({ hidden: false, stuck: false, animate: false })
      }
    } else {
      if (scrollY > height && !menu.open) {
        this.setState({ hidden: true })
      }
    }

    const unstickBufferZone = 15

    if (this.throttledScrollY > scrollY + unstickBufferZone && !this.state.stuck) {
      this.unstickAt = scrollY + unstickBufferZone
      this.setState({ stuck: true })
    } else if (scrollY > this.unstickAt && this.state.stuck && !menu.open) {
      this.setState({ stuck: false })
      delete this.unstickAt
    }

    if (this.lastScrollY > scrollY && this.state.stuck) {
      this.unstickAt = scrollY + unstickBufferZone
    }

    this.lastScrollY = scrollY
    this.sampleScrollSpeed()
  }

  sampleScrollSpeed = throttle(() => this.throttledScrollY = window.scrollY, 100)

}
