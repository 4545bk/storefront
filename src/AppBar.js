import React, { useContext } from 'react'
import { makeStyles } from '@material-ui/core/styles'
import { AppBar as MUIAppBar, Container, Toolbar, useScrollTrigger, Slide } from '@material-ui/core'
import PropTypes from 'prop-types'
import PWAContext from './PWAContext'

const useStyles = makeStyles(theme => ({
  /**
   * Styles applied to the root element.
   */
  root: {
    boxSizing: 'border-box',
    backgroundColor: theme.palette.background.default,
    boxShadow: 'none',
    borderBottom: `1px solid ${theme.palette.divider}`,
    zIndex: theme.zIndex.modal + 10,
  },
  /**
   * Styles applied to the spacer that fills the height behind the floating toolbar.
   */
  spacer: {
    boxSizing: 'border-box',
  },
  /**
   * Styles applied to the `Toolbar` element.
   */
  toolbar: {
    justifyContent: 'space-between',
    alignItems: 'center',
  },
  /**
   * Styles applied to the inner `Container` element.
   */
  container: {
    padding: 0,
  },
  /**
   * Styles applied to the offline warning element.
   */
  offline: {
    textAlign: 'center',
    backgroundColor: '#f34c4c',
    zIndex: 999999,
    width: '100vw',
    color: 'white',
  },
}))

export default function AppBar({ height, maxWidth, children, style, fixed, offlineWarning }) {
  const trigger = useScrollTrigger()
  const classes = useStyles()

  const { offline } = useContext(PWAContext)

  let appBar = (
    <MUIAppBar
      className={classes.root}
      style={{
        height,
        ...style,
      }}
    >
      <Container maxWidth={maxWidth} className={classes.container}>
        <Toolbar disableGutters className={classes.toolbar}>
          {children}
        </Toolbar>
      </Container>
    </MUIAppBar>
  )

  if (!fixed) {
    appBar = (
      <Slide appear={false} direction="down" in={!trigger}>
        {appBar}
      </Slide>
    )
  }

  return (
    <>
      <div style={{ height }} className={classes.spacer} />
      {offline && <div className={classes.offline}>{offlineWarning}</div>}
      {appBar}
    </>
  )
}

/**
 * The height of the AppBar
 */
AppBar.propTypes = {
  /**
   * Override or extend the styles applied to the component. See [CSS API](#css) below for more details.
   */
  classes: PropTypes.object,

  /**
   * The height of the AppBar in pixels
   */
  height: PropTypes.number,

  /**
   * The max width for the inner toolbar
   */
  maxWidth: PropTypes.oneOf(['xs', 'sm', 'md', 'lg', 'xl']),

  /**
   * Set as `true` if the AppBar should be fixed position.
   */
  fixed: PropTypes.bool,

  /**
   * String or Element to render within the offline warning container at the top of the app.
   */
  offlineWarning: PropTypes.oneOfType([PropTypes.string, PropTypes.element]),
}

AppBar.defaultProps = {
  height: 58,
  maxWidth: 'lg',
  offlineWarning: 'Your device lost its internet connection.',
  fixed: false,
}
