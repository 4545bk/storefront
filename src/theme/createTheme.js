import { createMuiTheme, adaptV4Theme } from '@mui/material/styles';
import merge from 'lodash/merge'

/**
 * Creates the default theme for your React Storefront app.  See Material UI's theme documentation
 * for more info: https://mui.com/customization/default-theme/
 * @param {Object} values
 * @return {Object} A material UI theme
 */
export default function createTheme(values = {}) {
  const theme = createMuiTheme(
    adaptV4Theme(merge(
      {},
      {
        zIndex: {
          modal: 999,
          amp: {
            modal: 2147483646,
          },
        },
        headerHeight: 64,
        loadMaskOffsetTop: 64 + 56 + 4,
        drawerWidth: 330,
        overrides: {},
      },
      values,
    )),
  )

  if (!theme.maxWidth) {
    theme.maxWidth = `${theme.breakpoints.values.lg}px`
  }

  return theme
}
