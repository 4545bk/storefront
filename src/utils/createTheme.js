import { createMuiTheme } from '@material-ui/core/styles'
import merge from 'lodash/merge'

export default function createTheme(values = {}) {
  let config = {}

  merge(
    config,
    {
      zIndex: {
        modal: 999,
        amp: {
          modal: 2147483646,
        },
      },
      headerHeight: 64,
      loadMaskOffsetTop: 64 + 56 + 4,
      drawerWidth: 330,
      overrides: {},
    },
    values,
  )

  const theme = createMuiTheme(config)

  if (!theme.maxWidth) {
    theme.maxWidth = `${theme.breakpoints.values.lg}px`
  }

  return theme
}
