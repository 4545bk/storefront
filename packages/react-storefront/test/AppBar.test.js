/**
 * @license
 * Copyright Â© 2017-2018 Moov Corporation.  All rights reserved.
 */
import React from 'react'
import { mount } from 'enzyme'
import AppBar from '../src/AppBar'
import createTheme from '../src/createTheme'
import { MuiThemeProvider } from '@material-ui/core'
import { Provider } from 'mobx-react'
import AppModelBase from '../src/model/AppModelBase'

describe("AppBar", () => {
  let theme

  beforeEach(function(){
    jest.spyOn(global.console, 'error').mockImplementation()
    theme = createTheme()
  })

  it('should render', () => {
    expect(mount(
      <Provider app={AppModelBase.create({})}>
        <MuiThemeProvider theme={theme}>
          <AppBar/>
        </MuiThemeProvider>
      </Provider>
    )).toMatchSnapshot()
  })
})