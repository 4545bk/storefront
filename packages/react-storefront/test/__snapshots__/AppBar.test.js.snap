// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppBar should render 1`] = `
<TestProvider>
  <Provider router={{...}} app={{...}} history={{...}}>
    <MuiThemeProviderOld theme={{...}}>
      <inject-WithStyles(Header)>
        <WithStyles(Header) menu={{...}} amp={false}>
          <Header menu={{...}} amp={false} classes={{...}} MenuIcon={[Function: WithStyles]} menuIconProps={{...}} menuAlign="left">
            <div className="RSFAppBar-root-1">
              <div className="RSFAppBar-wrap-4">
                <WithStyles(Toolbar) disableGutters={true} classes={{...}}>
                  <Toolbar disableGutters={true} classes={{...}} variant="regular">
                    <div className="MuiToolbar-root-11 RSFAppBar-toolBar-3 MuiToolbar-regular-13">
                      <Hidden mdUp={true} implementation="css" lgDown={false} lgUp={false} mdDown={false} smDown={false} smUp={false} xlDown={false} xlUp={false} xsDown={false} xsUp={false}>
                        <WithStyles(HiddenCss) mdUp={true} lgDown={false} lgUp={false} mdDown={false} smDown={false} smUp={false} xlDown={false} xlUp={false} xsDown={false} xsUp={false}>
                          <HiddenCss mdUp={true} lgDown={false} lgUp={false} mdDown={false} smDown={false} smUp={false} xlDown={false} xlUp={false} xsDown={false} xsUp={false} classes={{...}}>
                            <div className="MuiPrivateHiddenCss-mdUp-22">
                              <a on="tap:moov_menu.toggle" className="RSFAppBar-link-10">
                                <WithStyles(ToolbarButton) aria-label="Menu" color="inherit" onClick={[Function: res]} icon={{...}}>
                                  <ToolbarButton aria-label="Menu" color="inherit" onClick={[Function: res]} icon={{...}} classes={{...}}>
                                    <WithStyles(IconButton) classes={{...}} aria-label="Menu" color="inherit" onClick={[Function: res]}>
                                      <IconButton aria-label="Menu" color="inherit" onClick={[Function: res]} classes={{...}} disabled={false}>
                                        <WithStyles(ButtonBase) className="MuiIconButton-root-31 MuiIconButton-colorInherit-32" centerRipple={true} focusRipple={true} disabled={false} aria-label="Menu" onClick={[Function: res]}>
                                          <ButtonBase className="MuiIconButton-root-31 MuiIconButton-colorInherit-32" centerRipple={true} focusRipple={true} disabled={false} aria-label="Menu" onClick={[Function: res]} classes={{...}} component="button" disableRipple={false} disableTouchRipple={false} tabIndex="0" type="button">
                                            <button className="MuiButtonBase-root-37 MuiIconButton-root-31 MuiIconButton-colorInherit-32" onBlur={[Function]} onFocus={[Function]} onKeyDown={[Function]} onKeyUp={[Function]} onMouseDown={[Function]} onMouseLeave={[Function]} onMouseUp={[Function]} onTouchEnd={[Function]} onTouchMove={[Function]} onTouchStart={[Function]} onContextMenu={[Function]} tabIndex="0" type="button" disabled={false} aria-label="Menu" onClick={[Function: res]}>
                                              <span className="MuiIconButton-label-36">
                                                <div className="RSFToolbarButton-wrap-30">
                                                  <WithStyles(MenuIcon) open={false}>
                                                    <MenuIcon open={false} classes={{...}} label={true} CloseIcon={[Function: ShouldUpdate]}>
                                                      <div className="RSFMenuIcon-root-40">
                                                        <TransitionGroup component="div" childFactory={[Function: childFactory]}>
                                                          <div>
                                                            <CSSTransition classNames="rsf-menu-icon" timeout={300} onExited={[Function: bound bound handleExited]} in={true} appear={[undefined]} enter={[undefined]} exit={[undefined]}>
                                                              <Transition timeout={300} onExited={[Function]} in={true} appear={false} enter={true} exit={true} onEnter={[Function]} onEntered={[Function]} onEntering={[Function]} onExit={[Function]} onExiting={[Function]} mountOnEnter={false} unmountOnExit={false}>
                                                                <WithStyles(SvgIcon) className="RSFMenuIcon-icon-41">
                                                                  <SvgIcon className="RSFMenuIcon-icon-41" classes={{...}} color="inherit" component="svg" fontSize="default" viewBox="0 0 24 24">
                                                                    <svg className="MuiSvgIcon-root-42 RSFMenuIcon-icon-41" focusable="false" viewBox="0 0 24 24" color={[undefined]} aria-hidden="true" role="presentation">
                                                                      <OpenWithLabel>
                                                                        <svg version="1.1" id="Layer_1" x="0px" y="0px" viewBox="0 0 31.7 31.7">
                                                                          <path id="Shape" d="M0.9,17.7h29.8v-2.8H0.9V17.7L0.9,17.7z M0.9,10.8h29.8V8.1H0.9V10.8L0.9,10.8z M0.9,1.2v2.8h29.8V1.2H0.9 L0.9,1.2z" />
                                                                          <path id="menu" d="M0.4,24h1.1v1h0c0-0.1,0.1-0.2,0.2-0.3c0.1-0.1,0.3-0.3,0.4-0.4c0.2-0.1,0.4-0.2,0.6-0.3s0.5-0.1,0.8-0.1 c0.5,0,0.9,0.1,1.2,0.3c0.3,0.2,0.6,0.5,0.8,0.9c0.2-0.4,0.5-0.7,0.9-0.9s0.7-0.3,1.1-0.3c0.5,0,0.9,0.1,1.2,0.2 c0.3,0.2,0.6,0.4,0.7,0.6c0.2,0.3,0.3,0.6,0.4,0.9c0.1,0.3,0.1,0.7,0.1,1v3.7H8.7v-3.5c0-0.2,0-0.5,0-0.7c0-0.2-0.1-0.4-0.2-0.6 c-0.1-0.2-0.2-0.3-0.4-0.4S7.6,25,7.3,25c-0.6,0-1,0.2-1.2,0.5s-0.4,0.8-0.4,1.4v3.4H4.5v-3.3c0-0.3,0-0.6,0-0.8 c0-0.2-0.1-0.5-0.2-0.6c-0.1-0.2-0.2-0.3-0.4-0.4C3.7,25.1,3.5,25,3.2,25c-0.2,0-0.4,0-0.6,0.1c-0.2,0.1-0.4,0.2-0.5,0.4 c-0.2,0.2-0.3,0.4-0.4,0.6c-0.1,0.3-0.1,0.6-0.1,0.9v3.3H0.4V24z M12.3,27.6c0,0.3,0.1,0.5,0.2,0.8c0.1,0.2,0.3,0.4,0.5,0.6 c0.2,0.2,0.4,0.3,0.7,0.4c0.3,0.1,0.5,0.1,0.8,0.1c0.4,0,0.7-0.1,1-0.3s0.5-0.4,0.8-0.7l0.9,0.7c-0.7,0.9-1.6,1.3-2.8,1.3 c-0.5,0-0.9-0.1-1.3-0.3c-0.4-0.2-0.7-0.4-1-0.7c-0.3-0.3-0.5-0.6-0.6-1c-0.2-0.4-0.2-0.8-0.2-1.3c0-0.5,0.1-0.9,0.2-1.3 c0.2-0.4,0.4-0.8,0.7-1c0.3-0.3,0.6-0.5,1-0.7c0.4-0.2,0.8-0.3,1.3-0.3c0.6,0,1,0.1,1.4,0.3c0.4,0.2,0.7,0.4,1,0.8 c0.2,0.3,0.4,0.7,0.5,1.1c0.1,0.4,0.2,0.8,0.2,1.2v0.4H12.3z M16.1,26.7c0-0.3-0.1-0.5-0.1-0.7c-0.1-0.2-0.2-0.4-0.3-0.6 c-0.2-0.2-0.3-0.3-0.6-0.4c-0.2-0.1-0.5-0.1-0.8-0.1c-0.3,0-0.6,0.1-0.8,0.2c-0.2,0.1-0.4,0.3-0.6,0.4s-0.3,0.4-0.4,0.6 c-0.1,0.2-0.1,0.4-0.1,0.6H16.1z M18.6,24h1.2v1h0c0.2-0.3,0.4-0.6,0.8-0.8c0.4-0.2,0.8-0.3,1.3-0.3c0.3,0,0.6,0,0.9,0.1 c0.3,0.1,0.5,0.2,0.7,0.4c0.2,0.2,0.4,0.4,0.5,0.8c0.1,0.3,0.2,0.7,0.2,1.1v4.1H23v-3.7c0-0.3,0-0.5-0.1-0.8 c-0.1-0.2-0.2-0.4-0.3-0.5c-0.1-0.1-0.3-0.2-0.5-0.3C21.9,25,21.7,25,21.5,25c-0.2,0-0.5,0-0.7,0.1c-0.2,0.1-0.4,0.2-0.6,0.4 s-0.3,0.4-0.4,0.7c-0.1,0.3-0.1,0.6-0.1,0.9v3.2h-1.2V24z M31.3,30.3h-1.2v-1h0c-0.2,0.3-0.4,0.6-0.8,0.8c-0.4,0.2-0.8,0.3-1.3,0.3 c-0.3,0-0.6,0-0.9-0.1c-0.3-0.1-0.5-0.2-0.7-0.4c-0.2-0.2-0.4-0.4-0.5-0.8c-0.1-0.3-0.2-0.7-0.2-1.1V24h1.2v3.7c0,0.3,0,0.5,0.1,0.8 c0.1,0.2,0.2,0.4,0.3,0.5c0.1,0.1,0.3,0.2,0.5,0.3s0.4,0.1,0.5,0.1c0.2,0,0.5,0,0.7-0.1c0.2-0.1,0.4-0.2,0.6-0.4 c0.2-0.2,0.3-0.4,0.4-0.7s0.1-0.6,0.1-0.9V24h1.2V30.3z" />
                                                                        </svg>
                                                                      </OpenWithLabel>
                                                                    </svg>
                                                                  </SvgIcon>
                                                                </WithStyles(SvgIcon)>
                                                              </Transition>
                                                            </CSSTransition>
                                                          </div>
                                                        </TransitionGroup>
                                                      </div>
                                                    </MenuIcon>
                                                  </WithStyles(MenuIcon)>
                                                  <div />
                                                </div>
                                              </span>
                                              <NoSsr defer={false} fallback={{...}}>
                                                <WithStyles(TouchRipple) innerRef={[Function]} center={true}>
                                                  <TouchRipple center={true} classes={{...}}>
                                                    <TransitionGroup component="span" enter={true} exit={true} className="MuiTouchRipple-root-51" childFactory={[Function: childFactory]}>
                                                      <span className="MuiTouchRipple-root-51" />
                                                    </TransitionGroup>
                                                  </TouchRipple>
                                                </WithStyles(TouchRipple)>
                                              </NoSsr>
                                            </button>
                                          </ButtonBase>
                                        </WithStyles(ButtonBase)>
                                      </IconButton>
                                    </WithStyles(IconButton)>
                                  </ToolbarButton>
                                </WithStyles(ToolbarButton)>
                              </a>
                            </div>
                          </HiddenCss>
                        </WithStyles(HiddenCss)>
                      </Hidden>
                    </div>
                  </Toolbar>
                </WithStyles(Toolbar)>
              </div>
            </div>
          </Header>
        </WithStyles(Header)>
      </inject-WithStyles(Header)>
    </MuiThemeProviderOld>
  </Provider>
</TestProvider>
`;
